### PR TITLE
VP-5810: A new created order is not shown on the "Subscription.Orders" blade

### PR DESCRIPTION
**Pull request by AliveMen**
Clear subscription cache after order created
Add subscription validation

### CODE REVIEW

File: src/VirtoCommerce.SubscriptionModule.Core/Services/ISubscriptionService.cs [hunk 1]
```diff
@@ -12,5 +9,6 @@ public interface ISubscriptionService
         Task<Subscription[]> GetByIdsAsync(string[] subscriptionIds, string responseGroup = null);
         Task SaveSubscriptionsAsync(Subscription[] subscriptions);
         Task DeleteAsync(string[] ids);
+        Task<CustomerOrder> CreateOrderForSubscription(Subscription subscription);
```
Comment (Vectorfield4) [inline src/VirtoCommerce.SubscriptionModule.Core/Services/ISubscriptionService.cs:12]: From?

File: src/VirtoCommerce.SubscriptionModule.Data/Validation/SubscriptionValidator.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+using FluentValidation;
+using VirtoCommerce.SubscriptionModule.Core.Model;
+
+namespace VirtoCommerce.SubscriptionModule.Data.Validation
+{
+    public class SubscriptionValidator : AbstractValidator<Subscription>
+    {
+        public SubscriptionValidator()
+        {
+            RuleFor(subscription => subscription.StoreId).NotNull().NotEmpty();
```
Comment (Vectorfield4) [inline src/VirtoCommerce.SubscriptionModule.Data/Validation/SubscriptionValidator.cs]: One `NotEmpty()` is enough.
image

File: src/VirtoCommerce.SubscriptionModule.Data/Validation/SubscriptionValidator.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+            RuleFor(subscription => subscription.CustomerId).NotNull().NotEmpty();
```
Comment (Vectorfield4) [inline src/VirtoCommerce.SubscriptionModule.Data/Validation/SubscriptionValidator.cs]: Same.

File: tests/VirtoCommerce.SubscriptionModule.Test/SubscriptionExportImportTests.cs [hunk 4]
```diff
@@ -240,37 +202,23 @@ public async Task TestDataExport()
...
-            using (var resourceStream = ReadEmbeddedResource("Resources.SerializedSubscriptionData.json"))
-            {
-                await _subscriptionExportImport.DoImportAsync(resourceStream, IgnoreProgressInfo, _cancellationToken.Object);
-            }
+            var importStream = GetStreamFromString(actualJson);
+            await _subscriptionExportImport.DoImportAsync(importStream, IgnoreProgressInfo, _cancellationToken.Object);
 
-            // Assert
+            //Asset
             Assert.Equal(TestSubscriptions, actualSubscriptions);
```
Comment (yecli) [inline tests/VirtoCommerce.SubscriptionModule.Test/SubscriptionExportImportTests.cs]: You need to use another comparison here, which will compare all objects properties

**Review by Vectorfield4 — CHANGES_REQUESTED** [review]
Can you add tests cases for subscriptions (customerId=null,'empty'; storeId=null,'empty')?

File: tests/VirtoCommerce.SubscriptionModule.Test/SubscriptionExportImportTests.cs (hunk 4 above)
Comment (Vectorfield4) [reply tests/VirtoCommerce.SubscriptionModule.Test/SubscriptionExportImportTests.cs]: `TestSubscriptions.Should().BeEquivalentTo(actualSubscriptions);`

Comment (AliveMen) [reply tests/VirtoCommerce.SubscriptionModule.Test/SubscriptionExportImportTests.cs]: updated

File: src/VirtoCommerce.SubscriptionModule.Data/Validation/SubscriptionValidator.cs (hunk 2 above)
Comment (AliveMen) [reply src/VirtoCommerce.SubscriptionModule.Data/Validation/SubscriptionValidator.cs]: fixed

File: src/VirtoCommerce.SubscriptionModule.Data/Validation/SubscriptionValidator.cs (hunk 3 above)
Comment (AliveMen) [reply src/VirtoCommerce.SubscriptionModule.Data/Validation/SubscriptionValidator.cs]: fixed

File: src/VirtoCommerce.SubscriptionModule.Core/Services/ISubscriptionService.cs (hunk 1 above)
Comment (AliveMen) [reply src/VirtoCommerce.SubscriptionModule.Core/Services/ISubscriptionService.cs:12]: I think it is right. For subscription.

File: src/VirtoCommerce.SubscriptionModule.Data/Validation/SubscriptionValidator.cs (hunk 2 above)
Comment (AliveMen) [reply src/VirtoCommerce.SubscriptionModule.Data/Validation/SubscriptionValidator.cs]: fixed

File: src/VirtoCommerce.SubscriptionModule.Data/Validation/SubscriptionValidator.cs (hunk 3 above)
Comment (AliveMen) [reply src/VirtoCommerce.SubscriptionModule.Data/Validation/SubscriptionValidator.cs]: fixed

**Review by yecli — APPROVED** [review]

**Review by Vectorfield4 — APPROVED** [review]

File: src/VirtoCommerce.SubscriptionModule.Data/VirtoCommerce.SubscriptionModule.Data.csproj [hunk 5]
```diff
@@ -13,6 +13,7 @@
     <ProjectReference Include="..\VirtoCommerce.SubscriptionModule.Core\VirtoCommerce.SubscriptionModule.Core.csproj" />
   </ItemGroup>
   <ItemGroup>
+    <PackageReference Include="FluentValidation" Version="8.6.2" />
```
Comment (tatarincev) [inline src/VirtoCommerce.SubscriptionModule.Data/VirtoCommerce.SubscriptionModule.Data.csproj:16]: please check that this version doesn't higher than used in the platform or other modules, to avoid the side effect of an unexpected upgrade in the runtime.

**Review by tatarincev — APPROVED** [review]

File: src/VirtoCommerce.SubscriptionModule.Data/VirtoCommerce.SubscriptionModule.Data.csproj (hunk 5 above)
Comment (yecli) [reply src/VirtoCommerce.SubscriptionModule.Data/VirtoCommerce.SubscriptionModule.Data.csproj:16]: We are using 8.6.3 in the platform, so it is ok.
